Implements new reference syntax: array[index], list[index], map["key"…

adimova opened the pull request:
…], bean.field

Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

This PR is a part of the previous PR #5870.

It is possible that one module returns a complex data for Output value as array, list, map or bean.
To be able to fetch an internal value out, the follow-up module input reference must support the following syntax: array[index], list[index], map["key"], bean.field. Should be possible to chain references in one bigger expression. For example: {@code array[1].name.values}.

maggu2810 commented:
@adimova Can you please document the difference between #5995 and #6229 here?
I assume @triller-telekom would like to join the discussion.

adimova commented:
the difference is the syntax and it is documented - array[index], list[index], map["key"], bean.field

adimova commented:
Actually the big question here is "How to verify the connections?". We previously assumed (wrongly) that the Input/Output objects should provide this information. Also we simplified it to string and added tags in Input/Output objects for additional info. But in the common case, the module types are provided by developers and are more complex and generic. The real connection takes place between the modules and this fact brings us to the real need of the current change. But still we need to answer to the question "How to verify the connections?" or to not trying to do it in general.

triller-telekom commented:
I haven't had a look at the source of this PR yet, but what I understand from @adimova's explanation is that this PR here supports more than mine.

To me the the question is: How useful is a array[index] and list[index]? How do you ensure it is always the same sequence of items in there? What would be useful however, is to provide the possibility to hand over full arrays or full lists to another module, but not the x. th item of them...

So I think the map feature is useful as already discussed on my PR, because you have the key to access a specific element.

For the array[index] and list[index] I would like to see real use cases first, where we can discuss how useful they are.

adimova commented:
I give this opportunity to the user for completeness. The most useful capability of our automation module is that it allows pluggability. Every one can provide module handler implementation that provides an output - array or list and anyone can rely on such syntax as array[index] or list[index] to get a value from it. Do you say that you can predict that no one and ever will find this feature useful?
Except array[index] and list[index], my PR proposes also the getting a field from bean. Is this feature useful for you?

triller-telekom commented:
> Do you say that you can predict that no one and ever will find this feature useful?

I can't answer this *right now* and since you also do not know any use case for it, why do we want to implement it. Yes the automation engine is plug-able, extensible etc., *BUT* we should not over engineer it, i.e. only add features where we know that we need them *now*.

> my PR proposes also the getting a field from bean. Is this feature useful for you?

Could you please explain to me what this does? What is a "bean"? Where is it used and why do you want to get a field from it?

kaikreuzer commented:
> It is possible that one module returns a complex data for Output value as array, list, map or bean.

Well, technically this is possible, whether we want to allow it or not should probably be discussed and documented.

> Actually the big question here is "How to verify the connections?"

What would be your answer to it? Imho we should not give up the concept of being able to statically verify the connections, at least for the "standard" situations.

> We previously assumed (wrongly) that the Input/Output objects should provide this information.

I do not think that this is a wrong assumption, it just might not be applicable for "advanced" situations.

I personally would consider primitive types to be the standard and if those are used, we can build nice rule UIs that offer any match the right modules for a rule - this should be our ultimate goal for enabling user-friendly UIs).
Nonetheless we see that for more complex situations, especially the ones where developers want to pre-package templates with complex logic, we also want complex types as inputs/outputs and I guess for those cases we can waive the requirement to be able to statically validate connections. If something is incorrectly wired, it will only fail at runtime. But that's more or less the same situation as with scripts for rules, which are also "advanced" use cases (and not meant for the average user that wants to use a UI).

So I am in generell ok with that PR, I am only wondering, if we can restrict ourselves to certain complex types. Beans seem to make sense and we already use this feature in the event triggers, which output an "Event", which is a complex type. (@triller-telekom a "bean" should be considered a POJO, i.e. the framework can find getter methods for a given property.).
Maps are a "generic bean", which makes sense as well, if it isn't possible to specify the data structure statically as a bean at compile time already.
Lists are effectively a "multiple" and when allowing Maps, we probably won't get around Lists either.
Arrays are imho interchangeable with Lists. The issue that I see if we are supporting both is that if you want to consume the full complex object as an input, you will always have to check whether it is an Array or List and you have to deal with the conversion. Or, some output declares the type "Array" while another declares "List" as input, which renders both incompatible although they deal with the same structure. My preference would therefore be to drop Array support and rather define List. Wdyt?

triller-telekom commented:
> Nonetheless we see that for more complex situations, especially the ones where developers want to pre-package templates with complex logic, we also want complex types as inputs/outputs

If we consider the use of Lists as advance use-case, I am OK with

> Maps are a "generic bean", which makes sense as well, if it isn't possible to specify the data structure statically as a bean at compile time already.
Lists are effectively a "multiple" and when allowing Maps, we probably won't get around Lists either.

So Maps and Lists are OK with me then. @adimova WDYT?

htreu commented:
Hi @adimova, are you going to implement the suggested changes regarding arrays and lists any time soon?

kaikreuzer commented:
@adimova Short friendly ping!

htreu commented:
@adimova you may have noticed we are planing the 0.10.0 release. This PR should be part of it. Please comment with your plans. Thanks.

maggu2810 commented:
Would it make sense that @triller-telekom includes parts of this changes to his PR as @adimova seems to be not responding?

htreu commented:
Maybe @triller-telekom can take over this PR and finalize it? When I understand the last comment correct, Array and List handling has to be unified.

triller-telekom commented:
Currently I am rather busy with other things, so i probably won't be able to finish this in the near future. If anyone else would like to work on it, feel free. Sorry.

kaikreuzer commented:
I have updated this PR and removed the support of arrays in favour of lists.
This is imho now ready to be merged.

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/Connection.java:
```diff
@@ -12,64 +12,81 @@
  * defined by id of the module and name of the output.
  *
  * @author Yordan Mihaylov - Initial Contribution
+ * @author Ana Dimova - new reference syntax: array[index], list[index], map["key"], bean.field
```
remove `array[index]`

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/Connection.java:
```diff
@@ -12,64 +12,81 @@
-     * @param inputName     is an unique name of the {@code Input} in scope of the {@link Module}.
-     * @param ouputModuleId is an unique id of the {@code Module} in scope of the {@link Rule}.
-     * @param outputName    is an unique name of the {@code Output} in scope of the {@link Module}.
+     * @param reference
```
correct order and complete `reference`

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/Connection.java:
```diff
@@ -12,64 +12,81 @@
+     *
+     * @param inputName      is an unique name of the {@code Input} in scope of the {@link Module}.
+     * @param outputModuleId is an unique id of the {@code Module} in scope of the {@link Rule}.
+     * @param outputName     is an unique name of the {@code Output} in scope of the {@link Module}.
```
`reference` is missing in the JavaDoc

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/Connection.java:
```diff
@@ -78,34 +95,85 @@ public String getInputName() {
+            sb.append(outputName);
+        }
+        if (reference != null) {
+            sb.append(reference);
```
reference would be concatenated w/o a separator. Is this intended?

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/ConnectionValidator.java:
```diff
@@ -40,11 +41,15 @@
  * @author Ana Dimova - Initial contribution and API
  * @author Kai Kreuzer - refactored (managed) provider and registry implementation
  * @author Benedikt Niehues - validation of connection-types respects inheriting types
+ * @author Ana Dimova - new reference syntax: array[index], list[index], map["key"], bean.field
```
remove `array[index]`

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineImpl.java:
```diff
@@ -97,6 +97,7 @@
  * @author Kai Kreuzer - refactored (managed) provider, registry implementation and customized modules
  * @author Benedikt Niehues - change behavior for unregistering ModuleHandler
  * @author Markus Rathgeb - use a managed rule
+ * @author Ana Dimova - new reference syntax: array[index], list[index], map["key"], bean.field
```
same

kaikreuzer commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineImpl.java:
```diff
@@ -97,6 +97,7 @@
  * @author Kai Kreuzer - refactored (managed) provider, registry implementation and customized modules
  * @author Benedikt Niehues - change behavior for unregistering ModuleHandler
  * @author Markus Rathgeb - use a managed rule
+ * @author Ana Dimova - new reference syntax: array[index], list[index], map["key"], bean.field
```
thx, fixed.

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineImpl.java:
```diff
@@ -607,36 +603,34 @@ protected void postRuleStatusInfoEvent(String ruleUID, RuleStatusInfo statusInfo
-                } catch (Throwable t) {
+                } else {
                     if (sb == null) {
                         sb = new StringBuffer();
```
I guess StringBuilder is preferred to be used over StringBuilder.

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineImpl.java:
```diff
@@ -607,36 +603,34 @@ protected void postRuleStatusInfoEvent(String ruleUID, RuleStatusInfo statusInfo
                 }
+            } catch (Throwable t) {
+                if (sb == null) {
+                    sb = new StringBuffer();
```
same

htreu commented on bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineImpl.java:
```diff
@@ -589,9 +584,10 @@ protected void unsetEventPublisher(EventPublisher eventPublisher) {
      */
     protected void postRuleStatusInfoEvent(String ruleUID, RuleStatusInfo statusInfo) {
         if (eventPublisher != null) {
+            EventPublisher ep = eventPublisher;
```
why? also line 590.

kaikreuzer replied:
To avoid (valid) NPE compiler warnings. Field could be set to null while we are executing this code block.

kaikreuzer commented:
comments addressed

maggu2810 commented:
I don't think I can provide a "detailed" review fast enough (WRT the timeline).
If you tested it and it works, I would trust you and merge it (if it is okay for the others as well).

maggu2810 commented:
Travis takes its time...

kaikreuzer commented:
It finished!

maggu2810 commented:
Do you agree with that one:

```text
Implements new reference syntax: array[index], list[index], map["key"], bean.field

Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>
Also-by: Kai Kreuzer <kai@openhab.org>
```

kaikreuzer commented:
Yes - but the Also-By should be above the Signed-off-by as this has to be the last line in the message.
